src/Polly.Core/CircuitBreaker/Controller/CircuitStateController.cs

```diff
@@ -314,7 +337,16 @@ private void OpenCircuitFor_NeedsLock(Outcome<T> outcome, TimeSpan breakDuration
         scheduledTask = null;
         var utcNow = _timeProvider.GetUtcNow();
 
-        _blockedUntil = IsDateTimeOverflow(utcNow, breakDuration) ? DateTimeOffset.MaxValue : utcNow + breakDuration;
+        if (_breakDurationGenerator is not null && _behavior.FailureCount > 0)
+        {
+            var generatedBreakDuration = _breakDurationGenerator(_behavior.FailureCount);
+            _blockedUntil = IsDateTimeOverflow(utcNow, generatedBreakDuration) ? DateTimeOffset.MaxValue : utcNow + generatedBreakDuration;
+        }
+        else
+        {
+            _blockedUntil = IsDateTimeOverflow(utcNow, breakDuration) ? DateTimeOffset.MaxValue : utcNow + breakDuration;
+        }
+        
```

Comment:
This is less repetitive, but rather than mutating the value passed in it might be better to just compute the duration change in the caller instead.

```suggestion
        if (_breakDurationGenerator is not null && _behavior.FailureCount > 0)
        {
            breakDuration = _breakDurationGenerator(_behavior.FailureCount);
        }
        
        _blockedUntil = IsDateTimeOverflow(utcNow, breakDuration) ? DateTimeOffset.MaxValue : utcNow + breakDuration;
```